### PR TITLE
 improve v2 rate-limiter

### PR DESCRIPTION
**Pull request by MalloZup**
## Descriptions:

This PR improve the rate-limit interceptor  changing interface.

## Links:
This PR is follow-up of discussion at https://github.com/thanos-io/thanos/pull/3654

### CODE REVIEW

**Comment by MalloZup** [comment]
@googlebot I signed it!

:factory:

File: go.mod [hunk 1]
```diff
@@ -5,6 +5,7 @@ go 1.14
 require (
 	github.com/gogo/protobuf v1.2.1
 	github.com/golang/protobuf v1.3.2
+	github.com/juju/ratelimit v1.0.1
```
Comment (bwplotka) [inline go.mod:8]: Can we make this a... provider? 🤗

File: interceptors/ratelimit/examples_test.go [hunk 2]
```diff
@@ -2,32 +2,59 @@ package ratelimit_test
...
+	middleware "github.com/grpc-ecosystem/go-grpc-middleware/v2"
+	grpc_ratelimit "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/ratelimit"
+	"github.com/juju/ratelimit"
 	"google.golang.org/grpc"
+)
 
-	middleware "github.com/grpc-ecosystem/go-grpc-middleware/v2"
-	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/ratelimit"
+// Rationale:
+// we use in the rate-limiter interceptor a "tocken-bucket" algorithm
```
Comment (bwplotka) [inline interceptors/ratelimit/examples_test.go:14]: Rationale for what? Can we elaborate?

Also can we write comments full sentence?

File: interceptors/ratelimit/examples_test.go [hunk 3]
```diff
@@ -2,32 +2,59 @@ package ratelimit_test
...
+	rate = 5
+	// capacity of bucket. allow only 40 requests
+	tokenCapacity = 40
 )
 
-// alwaysPassLimiter is an example limiter which implements Limiter interface.
-// It does not limit any request because Limit function always returns false.
-type alwaysPassLimiter struct{}
+type rateLimiterInterceptor struct {
+	tokenBucket *ratelimit.Bucket
```
Comment (bwplotka) [inline interceptors/ratelimit/examples_test.go:28]: again I think this should land in provider similar to Prometheus and specific logging libs

File: interceptors/ratelimit/examples_test.go [hunk 4]
```diff
@@ -2,32 +2,59 @@ package ratelimit_test
...
+	tokenRes := r.tokenBucket.TakeAvailable(1)
+
+	// we reached rate limit, return an specific error for the clients
+	if tokenRes == 0 {
+		return fmt.Errorf("APP-XXX: reached Rate-Limiting %d", r.tokenBucket.Available())
+	}
 
-func (*alwaysPassLimiter) Limit(_ context.Context) bool {
-	return false
+	// we don't have reached rate-limit
```
Comment (bwplotka) [inline interceptors/ratelimit/examples_test.go:42]: https://thanos.io/tip/contributing/coding-style-guide.md/#comments-should-be-full-sentences

File: interceptors/ratelimit/examples_test.go [hunk 5]
```diff
@@ -2,32 +2,59 @@ package ratelimit_test
...
+	return nil
 }
 
 // Simple example of server initialization code.
 func Example() {
-	// Create unary/stream rateLimiters, based on token bucket here.
-	// You can implement your own ratelimiter for the interface.
-	limiter := &alwaysPassLimiter{}
+	// Create unary/stream rateLimiter
```
Comment (bwplotka) [inline interceptors/ratelimit/examples_test.go:48]: https://thanos.io/tip/contributing/coding-style-guide.md/#comments-should-be-full-sentences (missing trailing period)

.. and bellow

File: interceptors/ratelimit/ratelimit.go [hunk 6]
```diff
@@ -9,17 +9,17 @@ import (
 )
 
 // Limiter defines the interface to perform request rate limiting.
-// If Limit function return true, the request will be rejected.
+// If Limit function return an error, the request will be rejected.
```
Comment (bwplotka) [inline interceptors/ratelimit/ratelimit.go:12]: 💪🏽

Comment (bwplotka) [reply interceptors/ratelimit/ratelimit.go:12]:
```suggestion
// If Limit function returns an error, the request will be rejected with XXX status code and the provided error as the response.
```
We need to find solid XXX that makes sense I believe.

**Review by bwplotka — COMMENTED** [review]
LGTM, just some thoughts. Hope this makes sense (:

File: interceptors/ratelimit/examples_test.go (hunk 5 above)
Comment (MalloZup) [reply interceptors/ratelimit/examples_test.go:48]: thx will check it out. appreciated

**Comment by bwplotka** [comment]
Let me know when it's good for another review 🤗 

Fixes https://github.com/grpc-ecosystem/go-grpc-middleware/issues/351

**Comment by MalloZup** [comment]
@bwplotka  yop I will work this week. thx for feedback. I own you a cake or something :grin:

File: go.mod (hunk 1 above)
Comment (MalloZup) [reply go.mod:8]: yep make sense

**Comment by MalloZup** [comment]
@bwplotka  I do have reworked a bit the PR with providers.

 A part of the conflict file, I do have some doubts. I would appreciated if you have any feedback.

when Running `go test` locally I do have:
go test
```
# github.com/grpc-ecosystem/go-grpc-middleware/providers/tockenbucket/v2 [github.com/grpc-ecosystem/go-grpc-middleware/providers/tockenbucket/v2.test]
./examples_test.go:27:41: cannot use limiter (type TockenBucketInterceptor) as type "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/ratelimit".Limiter in argument to "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/ratelimit".UnaryServerInterceptor:
	TockenBucketInterceptor does not implement "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/ratelimit".Limiter (wrong type for Limit method)
		have Limit(context.Context) error
		want Limit() bool
./examples_test.go:30:41: cannot use limiter (type TockenBucketInterceptor) as type "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/ratelimit".Limiter in argument to "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/ratelimit".UnaryServerInterceptor:
	TockenBucketInterceptor does not implement "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/ratelimit".Limiter (wrong type for Limit method)
		have Limit(context.Context) error
		want Limit() bool
./examples_test.go:30:41: cannot use "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/ratelimit".UnaryServerInterceptor(limiter) (type grpc.UnaryServerInterceptor) as type grpc.StreamServerInterceptor in argument to middleware.WithStreamServerChain
FAIL	github.com/grpc-ecosystem/go-grpc-middleware/providers/tockenbucket/v2 [build failed]
```

I can't follow-up this error since I do have changed the interface to accept error. :thinking:

**Comment by bwplotka** [comment]
Yes. This is major problem with multi modules.

you need to play with `replace` to make sure it works for you. 

What you need is:
1. Change interface first. Merge PR to v2
1. Make a PR to add provider on top of new interface. Make sure to update code dep to latest.

Done (:

**Comment by MalloZup** [comment]
@bwplotka thx for confirmation, will play with that

**Comment by MalloZup** [comment]
@bwplotka take look on this. I have splitted the PR in 2 parts.

The 1st part , (this here) will just implement the `error` part.

The 2nd part PR follow-up will base on this and implement what I was doing previously

TIA :sunflower:

**Review by bwplotka — APPROVED** [review]
Solid.

Lgtm! Let's fix comment if you want (:

File: interceptors/ratelimit/ratelimit.go (hunk 6 above)
Comment (MalloZup) [reply interceptors/ratelimit/ratelimit.go:12]: I think I didn't saw that LOL. 
:grin:

**Comment by MalloZup** [comment]
@bwplotka  ok I missed the comment. should be fine now

File: interceptors/ratelimit/ratelimit.go [hunk 7]
```diff
@@ -9,17 +9,17 @@ import (
 )
 
 // Limiter defines the interface to perform request rate limiting.
-// If Limit function return true, the request will be rejected.
+// If Limit function returns an error, the request will be rejected with XXX status code and the provided error as the response.
```
Comment (bwplotka) [inline interceptors/ratelimit/ratelimit.go:12]: XXX? That was a placeholder ;p

Comment (MalloZup) [reply interceptors/ratelimit/ratelimit.go]: it sounded to me a bit suspicious actually :P ok removed :raccoon:

File: interceptors/ratelimit/ratelimit.go [hunk 8]
```diff
@@ -9,17 +9,17 @@ import (
 )
 
 // Limiter defines the interface to perform request rate limiting.
-// If Limit function return true, the request will be rejected.
+// If Limit function returns an error, the request will be rejected with status code and the provided error as the response.
```
Comment (bwplotka) [inline interceptors/ratelimit/ratelimit.go:12]:
```suggestion
// If Limit function returns an error, the request will be rejected with the gRPC codes.ResourceExhausted and the provided error.
```
I mean this (: Sorry for being not clear.

**Comment by MalloZup** [comment]
thx @bwplotka will follow-up with 2nd :heart:

**Pull request merged**